### PR TITLE
feat: Kafka Protocol

### LINKED ISSUE #196: Kafka protocol support
Are there any plans for Kafka protocol support in this sdk?

I would primarily be interested in using binary mode with Kafka, with Avro serialization for the data.  (Although structured mode for JSON would probably need to be supported, too.)

It seems the current HTTP CloudEvent does most of what is needed, except:

1. It needs a property to store the message Key.
2. For binary mode, the headers need to be prefixed with `ce_` (not `ce-`), with the exception of `content-type`.

I would not expect any specific serialization support, as I would rather be able to pass in any serializer of my choice such as the AvroSerializer from the confluent-kafka-python package, as a callable.

I'd be willing to help with this, but I'm not sure where to begin in this repo as it seems very http-centric at the moment.

### PR DESCRIPTION
**Pull request by davidwmartines**
Signed-off-by: davidwmartines <d5172@yahoo.com>

Fixes #196 

## Changes
This PR adds functions for converting CloudEvents to Kakfa-protocol-ready objects, and for converting Kafka-protocol messages into CloudEvents, per the [CloudEvents Kafka protocol binding](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/bindings/kafka-protocol-binding.md) spec.

 - Supports binary or structured mode conversions
 - Clients can supply appropriate serialization and deserializer functions as MarshallerType callables.

## Add Kafka Protocol support

- [x] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

**Comment by davidwmartines** [comment]
I will add tests.

File: cloudevents/tests/test_kafka_conversions.py [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+    return _deserialize
+
+
+def _throw():
+    raise Exception("fail")
+
+
+@pytest.fixture
+def bad_marshaller() -> types.MarshallerType:
+    return _throw
```
Comment (xSAVIKx) [inline cloudevents/tests/test_kafka_conversions.py:176]: why not just raise an error here?

File: cloudevents/tests/test_kafka_conversions.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+
+
```
Comment (xSAVIKx) [inline cloudevents/tests/test_kafka_conversions.py:178]: Can you please organize the tests created below under test classes? Probably one per method under a test.

File: cloudevents/tests/test_kafka_conversions.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+def test_to_structured_none_key(source_event):
+    source_event["key"] = None
+    result = to_structured(source_event)
+    assert result.key is None
+
+
+def test_to_structured_no_key(source_event):
+    del source_event["key"]
+    result = to_structured(source_event)
+    assert result.key is None
```
Comment (xSAVIKx) [inline cloudevents/tests/test_kafka_conversions.py:393]: I believe these tests should actually assert that an error is raised when no `partitionkey` is available in the event or the mapping function fails to do it's job.

**Review by xSAVIKx — COMMENTED** [review]
@davidwmartines thank you for your PR! This looks really good to me.

Can you please take a look at the comments posted below? I believe there is a bit of a misunderstanding between the spec and the implementation (or I may have misunderstood the spec). It'd be great to align the implementation with the spec or raise some questions regarding how the spec should translate into an implementation.

**Comment by davidwmartines** [comment]
@xSAVIKx Thanks for the great feedback.  I will get to work on this!

I agree on all points, except for requiring the key.  Kafka does allow for null message keys, so I think we can allow a key of `None`.   Let me know if you think this needs more discussion.  I will try to clarify this when I add docs to the KafkaMessage class.

Good call on the key mapper function -  I was wondering about that.  I will add the option to supply a key-mapper as well.

**Comment by davidwmartines** [comment]
working on key_mapper next...

File: cloudevents/kafka/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from cloudevents import exceptions as cloud_exceptions
```
Comment (xSAVIKx) [inline cloudevents/kafka/exceptions.py:1]: Can you please add a copyright header to this file?

**Review by xSAVIKx — APPROVED** [review]
@davidwmartines overall LGTM! thanks again for you work. Can you please address these 2 minor comments before we merge the PR?

**Comment by davidwmartines** [comment]
> @davidwmartines overall LGTM! thanks again for you work. Can you please address these 2 minor comments before we merge the PR?

Thanks!

**Review by xSAVIKx — APPROVED** [review]

**Pull request merged**